prov/gni/src/gnix_mbox_allocator.c

```diff
@@ -0,0 +1,537 @@
+			  struct gnix_slab *slab)
+{
+	size_t total_size;
+
+	if (!handle || !slab) {
+		GNIX_WARN(FI_LOG_EP_CTRL,
+			  "Invalid argument handle or slab.\n");
+		return -FI_EINVAL;
+	}
+
+	total_size = handle->page_size * __page_count(handle);
+
+	free_bitmap(slab->used);
+	free(slab->used);
+
+	GNI_MemDeregister(handle->nic_handle->gni_nic_hndl,
+			  &slab->memory_handle);
+
+	munmap(slab->base, total_size);
+
+	free(slab);
+
+	return FI_SUCCESS;
+}
+
+static int __find_free(struct gnix_mbox_alloc_handle *handle,
+		       struct gnix_slab **slab)
+{
+	struct slist_entry *entry;
+	struct gnix_slab *temp;
+	int ret;
+
+	for (entry = handle->slab_list.head; entry; entry = entry->next) {
+		temp = container_of(entry, struct gnix_slab, list_entry);
+		ret = find_first_zero_bit(temp->used);
+		if (ret >= 0)
+			break;
+	}
+
+	if (ret >= 0)
+		*slab = temp;
+
+	return ret;
+}
+
+static int __fill_mbox(struct gnix_mbox_alloc_handle *handle,
+		       struct gnix_slab *slab, size_t position,
+		       struct gnix_mbox **ptr)
+{
+	struct gnix_mbox *out;
+	size_t mapped_size;
+	int ret = FI_SUCCESS;
+
+	out = calloc(1, sizeof(*out));
+	if (!out) {
+		GNIX_WARN(FI_LOG_EP_CTRL,
+			  "Error allocating mbox: %s\n",
+			  strerror(errno));
+		ret = -FI_ENOMEM;
+		goto err;
+	}
+
+	mapped_size = handle->page_size * __page_count(handle);
+
+	out->slab = slab;
+	out->base = slab->base;
+	out->offset = (ret * handle->mbox_size);
+	out->memory_handle = &slab->memory_handle;
+
+	if (out->offset > mapped_size) {
+		GNIX_WARN(FI_LOG_EP_CTRL, "Mbox out of bounds.\n");
+		ret = -FI_EINVAL;
+		goto err1;
+	}
+
+	ret = test_and_set_bit(slab->used, position);
+	if (ret != 0) {
+		GNIX_WARN(FI_LOG_EP_CTRL,
+			  "Bit already set when creating mbox.\n");
+		ret = -FI_EINVAL;
+		goto err1;
+	}
+
+	*ptr = out;
+
+	return ret;
+
+err1:
+	free(out);
+err:
+	return ret;
+}
+
+int gnix_mbox_allocator_create(struct gnix_nic *nic, gni_cq_handle_t cq_handle,
+			       enum gnix_page_size page_size, size_t mbox_size,
+			       size_t mpmmap,
+			       struct gnix_mbox_alloc_handle **alloc_handle)
+{
+	struct gnix_mbox_alloc_handle *handle;
+	int ret;
+
+	handle = calloc(1, sizeof(*handle));
+	if (!handle) {
+		GNIX_WARN(FI_LOG_EP_CTRL,
+			  "Error allocating alloc handle: %s\n",
+			  strerror(errno));
+		ret = -FI_ENOMEM;
+		goto err;
+	}
+
+	handle->page_size = page_size * 1024 * 1024;
+	handle->mbox_size = mbox_size;
+	handle->mpmmap = mpmmap;
+	handle->nic_handle = nic;
+	handle->cq_handle = cq_handle;
+
+	ret = __open_huge_page(handle);
+	if (ret) {
+		GNIX_WARN(FI_LOG_EP_CTRL, "Error opening huge page.\n");
+		goto err1;
+	}
+
+	ret = __create_slab(handle);
+	if (ret) {
+		GNIX_WARN(FI_LOG_EP_CTRL, "Slab creation failed.\n");
+		goto err2;
+	}
+
+	*alloc_handle = handle;
+
+	return ret;
+
+err2:
+	free(handle->filename);
+err1:
+	free(handle);
+err:
+	return ret;
+}
+
+/*
+ * TODO: If one of the slabs still has memory assigned, should we still
```

Comment:
For a round #2 of this work, probably should return -FI_EBUSY if not all the mboxes have been returned to the slab.  But for this first checkin,  okay to leave as TODO.